File: 0931_Minimum_Falling_Path_Sum.cpp
class Solution {
public:
    int minFallingPathSum(vector<vector<int>>& matrix) {
        // set each cell to its minimum falling path sum
        for(int i = 1; i < matrix.size(); i++) {
            for(int j = 0; j < matrix[i].size(); j++) {
                int elem = matrix[i][j];
                matrix[i][j] = INT_MAX;
                
                for(int o : {-1, 0, 1}) {
                    if(j + o < 0 || j + o >= matrix[i - 1].size()) continue; // no such cell
                    matrix[i][j] = min(matrix[i][j], elem + matrix[i - 1][j + o]);
                }
            }
        }
        
        return *min_element(matrix[matrix.size() - 1].begin(), matrix[matrix.size() - 1].end());
    }
};

File: 0931_Minimum_Falling_Path_Sum.rs
use std::cmp::min; 

impl Solution {
    pub fn min_falling_path_sum(matrix: Vec<Vec<i32>>) -> i32 {
        let n = matrix.len();
        let m = matrix[0].len();
        // dp[i][j] = min sum of falling path to cell (i, j)
        let mut dp = matrix.clone();
        
        for i in 0..n {
            for j in 0..m {
                if i == 0 { continue; }
                
                let mut min_above = dp[i - 1][j];
                
                if j != 0 { min_above = min(min_above, dp[i - 1][j - 1]); }
                if j != m - 1 { min_above = min(min_above, dp[i - 1][j + 1]); }
                
                dp[i][j] += min_above;
            }
        }
        
        *dp[n - 1].iter().min().unwrap()
    }
}
